pit30m/data/log_reader.py

```diff
@@ -223,23 +254,50 @@ def get_lidar_geo_index(self, sort_by: str = "lidar_time") -> np.ndarray:
         index = index[np.argsort(index[sort_by])]
         return index
 
+    def filter_partitions(self, query_timestamps: np.ndarray, max_delta_s: float = 0.2) -> np.ndarray:
+        """Returns a boolean mask that indicates which sensor readings belong to the requested partitions."""
+
+        # Now, leave only the subset that matches the requested partition
+        partitions_mask = self._partitions_mask
+
+        # Associate with partititions mask
+        matched_indices, over = associate_np(query_timestamps, partitions_mask["timestamp"], max_delta_s=max_delta_s)
+        matched_bool = partitions_mask["mask"][matched_indices]
+        matched_bool[over] = False
+
+        # Filter with the obtained mask
+        return matched_bool
+
     @lru_cache(maxsize=16)
-    def get_cam_geo_index(self, cam_name: CamName, sort_by: str = "img_time") -> np.ndarray:
+    def get_cam_geo_index(
+        self,
+        cam_name: CamName = CamName.MIDDLE_FRONT_WIDE,
+        sort_by: str = "img_time",
+        max_delta_s=0.1,
+    ) -> np.ndarray:
         """Returns a camera index of dtype CAM_INDEX_V0_0_DTYPE.
         Args:
             cam_name: name of the camera index to load
             sort_by: name of the field that we want to sort by. Defaults to `img_time`
         Returns:
             A structured numpy array with the camera observations and their metadata.
         """
-        index_fpath = os.path.join(self.get_cam_root(cam_name), "index", f"index_v{self._index_version}.npz")
+
+        # Get the pre-computed index
+        # TODO(julieta) v0 and v1 have different name formatting on s3
+        index_fpath = os.path.join(self.get_cam_root(cam_name), "index", f"index_v{self._index_version:02d}.npz")
         if not self.fs.exists(index_fpath):
             raise ValueError(f"Index file not found: {index_fpath}!")
 
         with self.fs.open(index_fpath, "rb") as f:
             index = np.load(f)["index"]
 
         index = index[np.argsort(index[sort_by])]
+
+        # Filter by partitions
+        partitions_mask = self.filter_partitions(index["img_time"], max_delta_s)
```

Comment:
Are you sure this will be fast enough to always run on-demand whenever someone needs to read a log?

Reply:
This is surprisingly fast! And we can always make it faster with a custom implementation taking advantage of the fact that both arrays are sorted. I think the simplicity of being able to drop-in new indices or new partitions is well-worth it.

Reply:
Sounds awesome! Thanks for confirming!